src/coding_theory/hamming.lean

```diff
@@ -0,0 +1,234 @@
+end
+
+lemma eq_of_hamm_dist_eq_zero (x y : Π i, β i) : hamm_dist x y = 0 → x = y :=
+hamm_dist_eq_zero.mp
+
+lemma hamm_dist_ne_zero {x y : Π i, β i} : hamm_dist x y ≠ 0 ↔ x ≠ y :=
+not_iff_not.mpr hamm_dist_eq_zero
+
+lemma hamm_dist_pos {x y : Π i, β i} : 0 < hamm_dist x y ↔ x ≠ y :=
+by rw [←hamm_dist_ne_zero, iff_not_comm, not_lt, nat.le_zero_iff]
+
+@[simp] lemma hamm_dist_lt_one {x y : Π i, β i} : hamm_dist x y < 1 ↔ x = y :=
+by rw nat.lt_one_iff; exact hamm_dist_eq_zero
+
+lemma hamm_dist_eq_zero_iff_forall_eq {x y : Π i, β i} :
+  hamm_dist x y = 0 ↔ ∀ i, x i = y i :=
+by rw [hamm_dist_eq_zero, function.funext_iff]
+
+lemma hamm_dist_ne_zero_iff_exists_ne {x y : Π i, β i} :
+  hamm_dist x y ≠ 0 ↔ ∃ i, x i ≠ y i :=
+by rw [hamm_dist_ne_zero, function.ne_iff]
+
+section has_zero
+
+variable [Π i, has_zero (β i)]
+
+/-- The Hamming weight function to the naturals. -/
+
+def hamm_wt (x : Π i, β i) : ℕ := hamm_dist x 0
+
+lemma hamm_wt_eq (x : Π i, β i) : hamm_wt x = card {i // x i ≠ 0} := rfl
+
+lemma hamm_wt_eq_hamm_dist_zero (x : Π i, β i) : hamm_wt x = hamm_dist x 0 := rfl
+
+lemma hamm_wt_nonneg {x : Π i, β i} : 0 ≤ hamm_wt x := hamm_dist_nonneg
+
+lemma hamm_wt_le_card_fintype {x : Π i, β i} : hamm_wt x ≤ card ι := hamm_dist_le_card_fintype
+
+lemma hamm_wt_smul_le [has_zero α] [Π i, smul_with_zero α (β i)] (k : α)
+  (x : Π i, β i) : hamm_wt (k • x) ≤ hamm_wt x :=
+by simp_rw hamm_wt_eq_hamm_dist_zero; convert hamm_dist_smul_le; rw smul_zero'
+
+lemma hamm_wt_smul [has_zero α] [Π i, smul_with_zero α (β i)] {k : α}
+  (hk : ∀ i, is_smul_regular (β i) k) (x : Π i, β i) : hamm_wt (k • x) = hamm_wt x :=
+by convert hamm_dist_smul hk; rw smul_zero'; refl
+
+@[simp] lemma hamm_wt_eq_zero {x : Π i, β i} : hamm_wt x = 0 ↔ x = 0 := hamm_dist_eq_zero
+
+@[simp] lemma hamm_wt_zero : hamm_wt (0 : Π i, β i) = 0 := hamm_dist_self _
+
+lemma hamm_wt_ne_zero {x : Π i, β i} : hamm_wt x ≠ 0 ↔ x ≠ 0 := hamm_dist_ne_zero
+
+lemma hamm_wt_pos {x : Π i, β i} : 0 < hamm_wt x ↔ x ≠ 0 := hamm_dist_pos
+
+@[simp] lemma hamm_wt_lt_one {x : Π i, β i} : hamm_wt x < 1 ↔ x = 0 := hamm_dist_lt_one
+
+lemma hamm_wt_zero_iff_forall_zero {x : Π i, β i} : hamm_wt x = 0 ↔ ∀ i, x i = 0 :=
+hamm_dist_eq_zero_iff_forall_eq
+
+lemma hamm_wt_pos_iff_exists_nz {x : Π i, β i} : hamm_wt x ≠ 0 ↔ ∃ i, x i ≠ 0 :=
+hamm_dist_ne_zero_iff_exists_ne
+
+end has_zero
+
+lemma hamm_dist_eq_hamm_wt_sub [Π i, add_group (β i)] (x y : Π i, β i) :
+  hamm_dist x y = hamm_wt (x - y) :=
+by simp_rw [hamm_wt_eq, hamm_dist_eq, pi.sub_apply, sub_ne_zero]
+
+end
+
+section
+
+variables {α ι : Type*} {β : ι → Type*} [fintype ι] [Π i, decidable_eq (β i)]
+
+instance : has_dist (hamm β) := ⟨λ x y, hamm_dist (of_hamm x) (of_hamm y)⟩
+
+@[simp, push_cast] lemma dist_eq_hamm_dist (x y : hamm β) : dist x y =
+hamm_dist (of_hamm x) (of_hamm y) := rfl
+
+instance : pseudo_metric_space (hamm β) :=
+{ dist_self           :=  by push_cast; exact_mod_cast hamm_dist_self,
```

Comment:
Mathlib doesn't like using `;`s for single goals, as they behave badly in the infoview. This would usually instead be written
```suggestion
{ dist_self           := by { push_cast, exact_mod_cast hamm_dist_self },
```

Reply:
(this applies throughout)

Reply:
Oh... I was copying a style I'd seen in existing files. Happy to fix.